### PR TITLE
String based IO initialization for MRAA

### PR DESCRIPTION
**Pull request by mihaitghstefanescu**
Not very sure about the C++ part. It would probably be best to enclose calls over elements from the STL vectors with existing MRAA  C++ API calls. Will work on this asap.

### CODE REVIEW

File: src/initio/initio.c [hunk 1]
```diff
@@ -0,0 +1,794 @@
...
+  int mode = -1;
+  if (strncmp(proto[idx], S_MODE_0, strlen(S_MODE_0)) == 0) {
+    mode = MRAA_SPI_MODE0;
+  } else if (strncmp(proto[idx], S_MODE_1, strlen(S_MODE_1)) == 0) {
+    mode = MRAA_SPI_MODE1;
+  } else if (strncmp(proto[idx], S_MODE_2, strlen(S_MODE_2)) == 0) {
+    mode = MRAA_SPI_MODE2;
+  } else if (strncmp(proto[idx], S_MODE_3, strlen(S_MODE_3)) == 0) {
+    mode = MRAA_SPI_MODE3;
+  } else {}
```
Comment (alext-mkrs) [inline src/initio/initio.c]: Why the empty clause? Please remove if really unused. Same goes for all repeats below (I won't comment on all of them)

File: src/initio/initio.c [hunk 2]
```diff
@@ -0,0 +1,794 @@
...
+{
+  if (proto == NULL || n <= 1) {
+    syslog(LOG_ERR, "parse_i2c: invalid parameters");
+    return NULL;
+  }
+
+  int idx = 1;
+  mraa_i2c_context dev = NULL;
+
+  //i2c_bus, idx = 1
```
Comment (alext-mkrs) [inline src/initio/initio.c]: Please remove the unused code/comments. Same goes for repeats below.

File: src/initio/initio.c [hunk 3]
```diff
@@ -0,0 +1,794 @@
...
+    return MRAA_ERROR_NO_RESOURCES;
+  }
+
+  /* First, get rid of the leftover string. */
+  int mraa_upm_num_str = 0;
+  char **mraa_upm_str = mraa_tokenize_string(strdesc, MRAA_UPM_SEP, &mraa_upm_num_str);
+  if (mraa_upm_num_str == 0) {
+    syslog(LOG_ERR, "mraa_io_init: Failed to allocate memory for context");
+    free(new_desc);
+    return MRAA_ERROR_INVALID_PARAMETER;
```
Comment (alext-mkrs) [inline src/initio/initio.c]: If our error message just before this is "failed to allocate memory", then `MRAA_ERROR_NO_RESOURCES` is more appropriate here. However the error message looks wrong to me - if we've got 0 out of `mraa_tokenize_string()`, there's simply no such delimiter there, right? And that has nothgin to do with allocations. Please double-check.

File: src/initio/initio.c [hunk 4]
```diff
@@ -0,0 +1,794 @@
...
+        }
+      }
+    } else if (strncmp(str_tokens[0], IIO_KEY, strlen(IIO_KEY) ) == 0) {
+      mraa_iio_context dev = parse_iio(str_tokens, num_desc_tokens);
+      if (!dev) {
+        syslog(LOG_ERR, "mraa_io_init: error parsing iio");
+        status = MRAA_ERROR_INVALID_HANDLE;
+      }
+
+      // Allocate space for the new AIO context
```
Comment (alext-mkrs) [inline src/initio/initio.c]: `AIO` should be `IIO`

File: src/initio/initio.c [hunk 5]
```diff
@@ -0,0 +1,794 @@
...
+        }
+      }
+    } else if (strncmp(str_tokens[0], I2C_KEY, strlen(I2C_KEY) ) == 0) {
+      mraa_i2c_context dev = parse_i2c(str_tokens, num_desc_tokens);
+      if (!dev) {
+        syslog(LOG_ERR, "mraa_io_init: error parsing i2c");
+        status = MRAA_ERROR_INVALID_HANDLE;
+      }
+
+      // Allocate space for the new AIO context
```
Comment (alext-mkrs) [inline src/initio/initio.c]: `AIO` should be `I2C`

File: src/initio/initio.c [hunk 6]
```diff
@@ -0,0 +1,794 @@
...
+        }
+      }
+    } else if (strncmp(str_tokens[0], PWM_KEY, strlen(PWM_KEY) ) == 0) {
+      mraa_pwm_context dev = parse_pwm(str_tokens, num_desc_tokens);
+      if (!dev) {
+        syslog(LOG_ERR, "mraa_io_init: error parsing pwm");
+        status = MRAA_ERROR_INVALID_HANDLE;
+      }
+
+      // Allocate space for the new AIO context
```
Comment (alext-mkrs) [inline src/initio/initio.c]: `AIO` should be `PWM`

File: src/initio/initio.c [hunk 7]
```diff
@@ -0,0 +1,794 @@
...
+        }
+      }
+    } else if (strncmp(str_tokens[0], SPI_KEY, strlen(SPI_KEY) ) == 0) {
+      mraa_spi_context dev = parse_spi(str_tokens, num_desc_tokens);
+      if (!dev) {
+        syslog(LOG_ERR, "mraa_io_init: error parsing spi");
+        status = MRAA_ERROR_INVALID_HANDLE;
+      }
+
+      // Allocate space for the new AIO context
```
Comment (alext-mkrs) [inline src/initio/initio.c]: `AIO` should be `SPI`

File: src/initio/initio.c [hunk 8]
```diff
@@ -0,0 +1,794 @@
...
+        }
+      }
+    } else if (strncmp(str_tokens[0], UART_KEY, strlen(UART_KEY) ) == 0) {
+      mraa_uart_context dev = parse_uart(str_tokens, num_desc_tokens);
+      if (!dev) {
+        syslog(LOG_ERR, "mraa_io_init: error parsing uart");
+        status = MRAA_ERROR_INVALID_HANDLE;
+      }
+
+      // Allocate space for the new AIO context
```
Comment (alext-mkrs) [inline src/initio/initio.c]: `AIO` should be `UART`

File: src/initio/initio.c [hunk 9]
```diff
@@ -0,0 +1,794 @@
...
+        }
+      }
+    } else if (strncmp(str_tokens[0], UART_OW_KEY, strlen(UART_OW_KEY) ) == 0) {
+      mraa_uart_ow_context dev = parse_uart_ow(str_tokens, num_desc_tokens);
+      if (!dev) {
+        syslog(LOG_ERR, "mraa_io_init: error parsing uart_ow");
+        status = MRAA_ERROR_INVALID_HANDLE;
+      }
+
+      // Allocate space for the new AIO context
```
Comment (alext-mkrs) [inline src/initio/initio.c]: `AIO` should be `UART_OW`

File: api/mraa/initio.hpp [hunk 10]
```diff
@@ -0,0 +1,113 @@
+#pragma once
```
Comment (alext-mkrs) [inline api/mraa/initio.hpp]: License/copyright header is missing

File: examples/c++/initio.cpp [hunk 11]
```diff
@@ -0,0 +1,12 @@
+#include <iostream>
```
Comment (alext-mkrs) [inline examples/c++/initio.cpp]: License/copyright header is missing

File: examples/c/initio.c [hunk 12]
```diff
@@ -0,0 +1,24 @@
+#include <stdio.h>
```
Comment (alext-mkrs) [inline examples/c/initio.c]: License/copyright header is missing

File: include/initio/initio_keys.h [hunk 13]
```diff
@@ -0,0 +1,72 @@
+#pragma once
```
Comment (alext-mkrs) [inline include/initio/initio_keys.h]: License/copyright header is missing

File: src/initio/initio.c [hunk 14]
```diff
@@ -0,0 +1,781 @@
+#include <string.h>
```
Comment (alext-mkrs) [inline src/initio/initio.c]: License/copyright header is missing

File: src/initio/initio.c (hunk 1 above)
Comment (mihaitghstefanescu) [reply src/initio/initio.c]: Some Misra C rules dictate that. Will remove them

Comment (alext-mkrs) [reply src/initio/initio.c]: Yeah, I suspected that :) Makes no sense in mraa context, so yes, please.

**Comment by alext-mkrs** [comment]
@mihaitghstefanescu, I see you've added some commits, but the tests are still failing, is this ready for review?

File: src/initio/initio.c [hunk 15]
```diff
@@ -0,0 +1,848 @@
...
+    } else if (strncmp(proto[idx], G_EDGE_BOTH, strlen(G_EDGE_BOTH)) == 0) {
+        edge = MRAA_GPIO_EDGE_BOTH;
+    } else if (strncmp(proto[idx], G_EDGE_RISING, strlen(G_EDGE_RISING)) == 0) {
+        edge = MRAA_GPIO_EDGE_RISING;
+    } else if (strncmp(proto[idx], G_EDGE_FALLING, strlen(G_EDGE_FALLING)) == 0) {
+        edge = MRAA_GPIO_EDGE_FALLING;
+    }
+
+    if (edge != -1) {
+        if (mraa_gpio_dir(dev, (mraa_gpio_edge_t) edge) != MRAA_SUCCESS) {
```
Comment (pylbert) [inline src/initio/initio.c]: Double check here, this seems incorrect.

File: src/initio/initio.c [hunk 16]
```diff
@@ -0,0 +1,848 @@
...
+    /* Check for driver output mode. */
+    int driver_mode = -1;
+    if (strncmp(proto[idx], G_OUTPUT_OPEN_DRAIN, strlen(G_OUTPUT_OPEN_DRAIN)) == 0) {
+        driver_mode = MRAA_GPIO_OPEN_DRAIN;
+    } else if (strncmp(proto[idx], G_OUTPUT_PUSH_PULL, strlen(G_OUTPUT_PUSH_PULL)) == 0) {
+        driver_mode = MRAA_GPIO_PUSH_PULL;
+    }
+
+    if (driver_mode != -1) {
+        if (mraa_gpio_input_mode(dev, (mraa_gpio_out_driver_mode_t) driver_mode) != MRAA_SUCCESS) {
```
Comment (pylbert) [inline src/initio/initio.c]: This seems incorrect as well...

File: src/initio/initio.c (hunk 15 above)
Comment (mihaitghstefanescu) [reply src/initio/initio.c]: fixed

File: src/initio/initio.c (hunk 16 above)
Comment (mihaitghstefanescu) [reply src/initio/initio.c]: fixed

**Comment by mihaitghstefanescu** [comment]
Please guide me through the mock platform compatible pins I/O pins in order to successfully run our tests(which are commented at the current moment). I had some issues with them using the mock platform on Intel Joule. Thanks.

**Comment by pylbert** [comment]
Here are the resources mraa knows about with a mock platform build:

```
mraa-gpio list
00       GPIO0: GPIO
01        ADC0: AIO
02     I2C0SDA: I2C
03     I2C0SCL: I2C
04      SPI0CS: SPI
05    SPI0MOSI: SPI
06    SPI0MISO: SPI
07    SPI0SCLK: SPI
08     UART0RX: UART
09     UART0TX: UART
```

**Comment by pylbert** [comment]
@mihaitghstefanescu, try adding 

`-DBUILDARCH=MOCK`

to your cmake command.

File: src/initio/initio.c [hunk 17]
```diff
@@ -0,0 +1,854 @@
...
+            if (++idx == n)
+                return dev;
+        }
+    } else {
+        syslog(LOG_ERR, "parse_i2c: invalid i2c bus number");
+        return NULL;
+    }
+
+    int address = -1;
+    if (proto[idx] && (mraa_atoi_x(proto[idx], NULL, &address, 0) != MRAA_SUCCESS)) {
```
Comment (pylbert) [inline src/initio/initio.c]: should be 

== MRAA_SUCCESS

File: src/initio/initio.c [hunk 18]
```diff
@@ -0,0 +1,854 @@
...
+                return dev;
+            }
+        }
+    } else {
+        syslog(LOG_ERR, "parse_aio: invalid aio number");
+        return NULL;
+    }
+
+    int num_bits = -1;
+    if (proto[idx] && (mraa_atoi_x(proto[idx], NULL, &num_bits, 0) != MRAA_SUCCESS)) {
```
Comment (pylbert) [inline src/initio/initio.c]: should be 

== MRAA_SUCCESS

**Comment by pylbert** [comment]
@mihaitghstefanescu, looking good so for.

Can we get this rebased onto master?  

Also, it may help to squash some of these commits into logical groups.

File: api/mraa/initio.hpp (hunk 10 above)
Comment (alext-mkrs) [reply api/mraa/initio.hpp]: Fixed now

File: include/initio/initio_keys.h [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+ */
+
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/*
+ *NOTE: GPIO KEYS
```
[the remark this answers is not in the document]
Comment (mihaitghstefanescu) [reply include/initio/initio_keys.h]: The file has already been run through clang-format. I have also removed the notes.

File: src/initio/initio.c [hunk 20]
```diff
@@ -0,0 +1,854 @@
...
+{
+    if (proto == NULL || n <= 1) {
+        syslog(LOG_ERR, "parse_i2c: invalid parameters");
+        return NULL;
+    }
+
+    int idx = 1;
+    mraa_i2c_context dev = NULL;
+
+    // i2c_bus, idx = 1
```
[the remark this answers is not in the document]
Comment (mihaitghstefanescu) [reply src/initio/initio.c]: Fixed.

File: src/initio/initio.c [hunk 21]
```diff
@@ -0,0 +1,854 @@
...
+                return dev;
+            }
+        }
+    } else {
+        syslog(LOG_ERR, "parse_uart: invalid uart baudrate");
+    }
+
+    /* Check for mode - [int] bytesize, [mraa_uart_parity_t] parity, [int] stopbits*/
+    char* end = proto[idx];
+    /* Check for bytesize*/
```
[the remark this answers is not in the document]
Comment (mihaitghstefanescu) [reply src/initio/initio.c]: Fixed.

File: src/initio/initio.c [hunk 22]
```diff
@@ -0,0 +1,854 @@
...
+        } else {
+            if (++idx == n) {
+                return dev;
+            }
+        }
+    } else {
+        syslog(LOG_ERR, "parse_uart: invalid uart baudrate");
+    }
+
+    /* Check for mode - [int] bytesize, [mraa_uart_parity_t] parity, [int] stopbits*/
```
[the remark this answers is not in the document]
Comment (mihaitghstefanescu) [reply src/initio/initio.c]: Fixed.

File: api/mraa/initio.hpp [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+        } else {
+            leftoverStr = std::string("");
+        }
+    }
+
+    MraaIo() : descs() {}
+
+    ~MraaIo()
+    {
+        if (descs->leftover_str)
```
[the remark this answers is not in the document]
Comment (mihaitghstefanescu) [reply api/mraa/initio.hpp]: Fixed.

File: include/initio/initio_keys.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (alext-mkrs) [reply include/initio/initio_keys.h]: @mihaitghstefanescu that's interesting, probably it's not able to format those or that's not in our config file... Ok, that's for correcting - but now you have spacing inconsistency again. In those comments that replaced those with notes, there's no space after the first `*` and later in the file, there is. E.g. [here](https://github.com/intel-iot-devkit/mraa/pull/899/files#diff-66ec42367a1caabd47c72a1be64ceee0R40) it's inconsistent

**Comment by alext-mkrs** [comment]
Ok, LGTM now with a small nitpick on comment spacing in `initio_keys.h` (see comment above). Provided that's fixed, I'm good with merging.

**Comment by alext-mkrs** [comment]
Oh, and assuming the test failures I see in Travis are fixed too, naturally.

File: src/initio/initio.c [hunk 24]
```diff
@@ -0,0 +1,852 @@
...
+#include "mraa_internal.h"
+
+#define DESC_SEP ","
+#define TOK_SEP ":"
+
+static mraa_result_t
+mraa_atoi_x(const char* intStr, char** str_end, int* value, int base)
+{
+    long val = strtol(intStr, str_end, base);
+    if (errno == ERANGE || val > INT_MAX || val < INT_MIN) {
```
Comment (pylbert) [inline src/initio/initio.c]: We should also check for the case where no conversion happens.  E.g. val = 0 and intStr == str_end and fail.

File: src/initio/initio.c [hunk 25]
```diff
@@ -0,0 +1,852 @@
...
+    }
+
+    int idx = 1;
+    mraa_aio_context dev = NULL;
+
+    int aio_num = -1;
+    if (proto[idx] && (mraa_atoi_x(proto[idx], NULL, &aio_num, 0) == MRAA_SUCCESS)) {
+        dev = mraa_aio_init(aio_num);
+        if (dev == NULL) {
+            syslog(LOG_ERR, "parse_aio: could not init aio number %d", aio_num);
```
Comment (pylbert) [inline src/initio/initio.c]: Ideally, all errors would tell the user exactly what a failure is and give them a hint on how to solve it.  

For example:

Could not initialize aio '1' from initialization string 'a:1:out'

File: src/initio/initio.c [hunk 26]
```diff
@@ -0,0 +1,852 @@
...
+            return NULL;
+        } else {
+            if (++idx == n) {
+                return dev;
+            }
+        }
+    } else {
+        syslog(LOG_ERR, "parse_aio: invalid aio number");
```
Comment (pylbert) [inline src/initio/initio.c]: and

Could not convert '1blahblah' to integer from initialization string 'a:1blahblah:out'

**Comment by pylbert** [comment]
Nice work @mihaitghstefanescu!  Merged.

**Pull request closed without merging**